### ISSUE TITLE
Infinite execution on some input

### ISSUE THREAD

**Issue by dattaz** [Python-Markdown/markdown#551]
With some input, i have a infinite execution, with markdown function, no exception raise.

Step to reproduce: https://gist.github.com/anonymous/ffab9ad433127893f04b9d009cd21444

**Comment by waylan** [comment]
Thanks for the report.

First from a practical standpoint, why is that not in a code block or span? That is not Markdown code but is close enough collisions would not be surprising. As it turns out, most implementations just pass it through unaltered, although there are some variations. Interestingly, at least three other implementations fail to parse it at all (according to Babelmark [^1]).

That said, you shouldn't be able to crash a server by submitting some plain text, so we do need to fix this. However, the fix will be to simply have it not crash/enter an infinite loop. What is actually returned is unpredictable as this is not really "valid" Markdown.

Without doing any debugging, my best guess is that the nested brackets regex (for the link syntax) is the problem. Unlike in Perl (which the reference implementation was written in), Python's regex engine doesn't do recursion so we fake it with six levels of nested hard-coded in. The assumption was that no one would ever try to create more than that. Personally, I was never crazy about the solution, but it was in-place long before I got involved and nothing better has ever been proposed.

[^1]: I did not include the Babelmark link as it causes my server to freeze and I have to manually reset it. Please don't copy the input to Babelmark yourself. Thank you.

**Comment by dattaz** [comment]
I'm totally agree with you, it's should be code block, but it's data from stackoverflow dump that hasn't been format in the good way by the author of the comment.

> However, the fix will be to simply have it not crash/enter an infinite loop. What is actually returned is unpredictable as this is not really "valid" Markdown

Yes it's logical, because main problem is infinite loop

**Comment by waylan** [comment]
So it does not appear to be an infinite loop. If you wait long enough it will finish, but most servers would time out by then. And I seem to have narrowed the minimum input down to this:

```
[[&quot;][&quot;][&quot;][&quot;][&quot;][&quot;][&quot;][&quot;][&quot;]]
```

As you remove each `[&quot;]` from the string, it runs noticeably faster. In order to replicate it, the opening and closing double brackets are necessary as is having an HTML entity in each set of brackets.

**Comment by waylan** [comment]
Interestingly, this seemed to not be a noticeable problem even though it contains the same number of sub-strings:

```
[[&quot;][&quot;][&quot;][&quot;][&quot;][&quot;]] [[&quot;][&quot;][&quot;][&quot;][&quot;][&quot;]]
```

I needed at least 8 sub-strings within one parent set of brackets before the problem became apparent by me on my machine.

**Comment by Nicals** [comment]
Here is what I got so far. The tests where performed with the following markdown content.

```python
import markdown
markdown.markdown('[[&quot;][&quot;][&quot;][&quot;][&quot;][&quot;][&quot;][&quot;][&quot;]]')
```

For a 12 second markdown conversion, the callgraph is:

```
12.746 <module>  perf.py:3
└─ 12.731 markdown  markdown/__init__.py:478
   └─ 12.727 convert  markdown/__init__.py:332
         └─ 12.725 run  markdown/treeprocessors.py:264
                  └─ 12.725 __handleInline  markdown/treeprocessors.py:88
                              └─ 12.725 __applyPattern  markdown/treeprocessors.py:218
```

So all the time is spent in markdown.treeprocessors.InlineProcessor::\_\_apply\_pattern.

All the time is spend in the *math* method of *_sre.SRE_Pattern.match* (99.59% of the time, see prof file attached).
[prof.zip](https://github.com/waylan/Python-Markdown/files/827136/prof.zip)

The match that is slow is at [markdown.teeprocessors:233][TREEPROC_LINE].
I did a quick and really dirty *profiling* (if you allow me calling it that way) that gaves me interresting results:

```
  218     def __applyPattern(self, pattern, data, patternIndex, startIndex=0):                                                                         
+ 233         import time                                                              
+ 234         start_time = time.time()                                                 
+ 235         print data                                                               
+ 236         print data[startIndex:]                                                  
+ 237         print pattern                                                            
+ 238         print pattern.compiled_re.pattern                                        
  239         match = pattern.getCompiledRegExp().match(data[startIndex:])                                                                      
+ 240         print time.time() - start_time                                           
+ 241         print '-' * 40                                                           
  242         leftData = data[:startIndex]
```

```shell-session
$ time python perf.py > dirty_prof

real	0m14,001s
user	0m13,992s
sys	0m0,012s
(python-markdown)
```

```
[[&quot;][&quot;][&quot;][&quot;][&quot;][&quot;][&quot;][&quot;][&quot;]]
[[&quot;][&quot;][&quot;][&quot;][&quot;][&quot;][&quot;][&quot;][&quot;]]
<markdown.inlinepatterns.ReferencePattern object at 0x7f6162ec0c10>
^(.*?)(?<!\!)\[([^\]\[]*(\[[^\]\[]*(\[[^\]\[]*(\[[^\]\[]*(\[[^\]\[]*(\[[^\]\[]*(\[[^\]\[]*\])*[^\]\[]*\])*[^\]\[]*\])*[^\]\[]*\])*[^\]\[]*\])*[^\]\[]*\])*[^\]\[]*)\]\s?\[([^\]]*)\](.*)$
7.16163802147
----------------------------------------
[[&quot;][&quot;][&quot;][&quot;][&quot;][&quot;][&quot;][&quot;][&quot;]]
[[&quot;][&quot;][&quot;][&quot;][&quot;][&quot;][&quot;][&quot;][&quot;]]
<markdown.inlinepatterns.LinkPattern object at 0x7f6162ec0c50>
^(.*?)(?<!\!)\[([^\]\[]*(\[[^\]\[]*(\[[^\]\[]*(\[[^\]\[]*(\[[^\]\[]*(\[[^\]\[]*(\[[^\]\[]*\])*[^\]\[]*\])*[^\]\[]*\])*[^\]\[]*\])*[^\]\[]*\])*[^\]\[]*\])*[^\]\[]*)\]\(\s*(<.*?>|((?:(?:\(.*?\))|[^\(\)]))*?)\s*((['"])(.*?)\12\s*)?\)(.*)$
6.77018499374
----------------------------------------
```

There is a total of 40 matches calls. I only shows the 2 that are really slow.
The full generated file is here:
[dirty_prof.txt](https://github.com/waylan/Python-Markdown/files/827148/dirty_prof.txt)

That's all for today. I hope it will help. Seems we have some regex review to do...

[TREEPROC_LINE]: https://github.com/waylan/Python-Markdown/blob/master/markdown/treeprocessors.py#L233

**Comment by waylan** [comment]
>     <markdown.inlinepatterns.ReferencePattern object at 0x7f6162ec0c10>
>     ^(.*?)(?<!\!)\[([^\]\[]*(\[[^\]\[]*(\[[^\]\[]*(\[[^\]\[]*(\[[^\]\[]*(\[[^\]\[]*(\[[^\]\[]*\])*[^\]\[]*\])*[^\]\[]*\])*[^\]\[]*\])*[^\]\[]*\])*[^\]\[]*\])*[^\]\[]*)\]\s?\[([^\]]*)\](.*)$

>     <markdown.inlinepatterns.LinkPattern object at 0x7f6162ec0c50>
>     ^(.*?)(?<!\!)\[([^\]\[]*(\[[^\]\[]*(\[[^\]\[]*(\[[^\]\[]*(\[[^\]\[]*(\[[^\]\[]*(\[[^\]\[]*\])*[^\]\[]*\])*[^\]\[]*\])*[^\]\[]*\])*[^\]\[]*\])*[^\]\[]*\])*[^\]\[]*)\]\(\s*(<.*?>|((?:(?:\(.*?\))|[^\(\)]))*?)\s*((['"])(.*?)\12\s*)?\)(.*)$

@Nicals that is helpful. It confirms what I already suspected. 

> Without doing any debugging, my best guess is that the nested brackets regex (for the link syntax) is the problem. Unlike in Perl (which the reference implementation was written in), Python's regex engine doesn't do recursion so we fake it with six levels of nested hard-coded in. The assumption was that no one would ever try to create more than that. Personally, I was never crazy about the solution, but it was in-place long before I got involved and nothing better has ever been proposed.

The regex for nested brackets (see [LINK_RE][1] and [REFERENCE_RE][3]) is clearly the problem (see also [IMAGE_LINK_RE][2] and [IMAGE_REFERENCE_RE][4] which share the same regex but are not being run here).

[1]: https://github.com/waylan/Python-Markdown/blob/master/markdown/inlinepatterns.py#L130
[2]: https://github.com/waylan/Python-Markdown/blob/master/markdown/inlinepatterns.py#L134
[3]: https://github.com/waylan/Python-Markdown/blob/master/markdown/inlinepatterns.py#L137
[4]: https://github.com/waylan/Python-Markdown/blob/master/markdown/inlinepatterns.py#L143

**Comment by facelessuser** [comment]
I think it is possible this and #613 could be fixed by switching to a more capable regular expression engine, specifically [regex](https://pypi.python.org/pypi/regex/).

The regex module allows for recursive regular expression.  And while recursion should be used sparingly, I think this is a good case for it.

A simple example is using an [atomic group](https://www.regular-expressions.info/atomic.html) `(?>)` and we recursively evaluate the square brackets by referencing the first group with `(?1)`:

```
(?<!\!)(\[(?>[^\[\]]|(?1))*\])\s?\[([^\]]*)\]
```

The regex module should be used probably with version0 flag (the default) which should be compatible with Re patterns.  Not sure about performance yet, so if performance is just as bad, then it won't matter.  So, obviously this would need some good testing before hand, but I think it would clear up the limitations of the link patterns, and hopefully perform better, but all of this remains to bee seen.  I'll probably run some experiments to test compatibility and performance, but I wanted to document the idea here before I start.

I'd like to think this could be a drop in fix with no side effects, if it actually fixes performance.  But, if it breaks compatibility, maybe it would be acceptable for 3.0 release, but I'm hoping it could work for 2.6.X.

Anyways, just an untested idea at this point.

**Comment by waylan** [comment]
> I think it is possible this and #613 could be fixed by switching to a more capable regular expression engine, specifically regex.

Generally, I'm not opposed to the idea. However, it appears that the lib includes a C extension, which may or may not be problematic for some to install. Windows specific wheels are available for download on PyPI (which eliminates the need for Windows users to compile), but only the source `tar.gz` is provided for other platforms (which presumably requires a compile on install). Some shared web hosts I've used in the past (pre-pip days) allowed users to install pure Python packages, but not anything that needed to compile C extensions (admittedly less of a concern with pip/setuptools and Python's support for user specific site-packages directories today).  However, I've also had local Debian/Ubuntu systems which wouldn't compile C extensions without the python-dev package installed (not the default). I'd hate to cut off some users or create extra installation hurdles due to a third party dependency issue. This may or may not be an issue in practice. Just noting another potential point of concern that needs tested/reviewed.

**Comment by facelessuser** [comment]
> However, I've also had local Debian/Ubuntu systems which wouldn't compile C extensions without the python-dev package installed (not the default).

Yes, this may or may not be an issue, and ultimately, some decision will have to be made if this is an issue; that's the kind of stuff I leave to you 🙂.  I'll be happy to test in anyway I can to give us an idea of the kind of impact it might have.  Currently, I'm more just looking at possible solutions.  Once we know what does and does not fix it, we can at least make an educated decision of applying a fix, partial fix, or can't/won't fix.

Maybe using an re pattern that does the 6 levels of recursion without capturing groups would speed things up, I'm not sure, but that is also something I want to look at as well. Currently we are using capturing groups which we don't need.  But using non-capturing groups would still be limited to 6 levels, which is the other issue I was hoping to address, but if that is a necessary evil, then that is fine.

Unfortunately, the problem we are looking may not be easily solvable with `re`.  Maybe `regex` can handle it better, but if it can't, the only other solutions that I can think of involves parsing the text without regular expressions which I don't think fits with the current framework.

**Comment by facelessuser** [comment]
Some results:

1. Changing to non-capturing groups with re still hangs for a long time.

2. Using regex with recursive pattern completed instantly.  In this test, I only forced tests derived from `Pattern` to compile with `regex`, so I did not do a library wide force to regex from re.  My aim in this initial test was only to test if this would fix the above issue, which it did. And make sure that tests still passed.

Next step would be to change all usage of re to regex and ensure tests still passed.  

I'll post a test branch at some point.  I can't see another way around this except to move to `regex` or to include non regular expression plugins.  Once this effort is complete, we can discuss whether we want to risk including a dependency that includes C code.

**Comment by waylan** [comment]
I wonder if this is related to #161. At the very least it would be interesting to see how switching to `regex` affects [this test][1]. For that matter, I suspect this issue may actually be a subset of that issue.

> I can't see another way around this except to move to `regex` or to include non regular expression plugins

I have entertained switching to some sort of token parser which would generate a stream of tokens. The idea being that rather than running a bunch of regexs in a particular order, each element is parsed in the order it appears in the document ([here][2] is a rudimentary experiment I started and never followed up on). Of course, that would completely break backward compatibility for existing extensions... unless we implemented two types of inline parsers, the existing regex based and a new one which most/all of the builtin patterns could be converted to. 

[1]: https://github.com/Python-Markdown/markdown/issues/161#issuecomment-10813311
[2]: https://gist.github.com/waylan/273e4a67febfde120257

**Comment by facelessuser** [comment]
> I wonder if this is related to #161. At the very least it would be interesting to see how switching to regex affects this test. For that matter, I suspect this issue may actually be a subset of that issue.

I'm sure this is contributing to the problem.  I suspect you may have been on the right track with:

> `"^(.*?)%s(.*?)$" % pattern`

I can't think of any plugin that uses group 1 and group -1 to do anything.  I've often wondered why those are included.  I feel like throwing those out would probably help. Every returned match is going to return from the starting index to the start of the match, and then from the end of the match to the end of the paragraph.  On large paragraphs, that seems like we'd be taking a huge performance hit.

> I have entertained switching to some sort of token parser which would generate a stream of tokens.

Without looking at other Markdown source, I can almost guarantee there are parsers that do exactly this.  I feel like it is easier to parse actual markdown this way as you deal with your specific tokens as you encounter them.  I think it is harder to customize in this approach as well which is also why I'm not against the fact that Python Markdown uses regular expression.  I like the fact that Python Markdown is so extendable.  I don't think `re` is fantastic regular expression engine, but it is much better than JavaScript's though 🙂.

I think something like `regex` brings modern regular expression into Python, I personally use `regex` in my own [search and replace tool](https://github.com/facelessuser/Rummage) I wrote.  But I also see why someone may be hesitant to bring in `regex` if they have an established user base that expect a pure Python solution.  I know `regex` doesn't have extreme build requirements, but I'm sure there are probably some installation hurdles in specific cases.

---

If I had to pick out Python Markdown's key flaws, they are:

1. Inefficient inline parsing: `"^(.*?)%s(.*?)$" % pattern` seems like an issue which could easily be fixed, but it would not be backwards compatible.  Using `search` would probably speed things up, but only if you drop `^(.*?)` and `(.*?)$`, but I don't think `match` inherently causes an issue.  And if you could iterate over the paragraph once, it might be even better.

    I don't recall how Python Markdown does things, maybe it already does this.  But I guess running a match pattern at the current char until something handles it, and then setting the new new index to the end of that match would be best.  If nothing matches, you increment to the next char.  All the unhandled chars would get added to the tail of the previous element or appended to the current text of the parent element. As you generate tags, you would recursively pass them through as well. Anyways, I'm probably oversimplifying, and maybe that is exactly what it is doing already, but I think the real problem is `^(.*?)` and `(.*?)$`.

2. How Python Markdown naively handles blocks.

    Digging back in the code history, I see that the library once had line plugins, and sometimes I think line plugins would actually make handling blocks syntax easier.  If you are talking about simple paragraphs, then the way we do blocks makes sense, but we have all of these other blocks that don't fit in that box.

    Nested indented code blocks lose empty lines because we naively split on `\n\n`. A matter of fact, any block that contains `\n\n` is a pain to process.  You almost have to resort to preprocessing which should only be used for things like whitespace or preprocessing stuff like CriticMarkup.

    I think HTML and fence parsing would kind of work themselves out if block handling was done better or abandoned for something like line handling.  Some blocks need the ability to scan ahead, and in nested blocks we often loose the ability to see see further ahead.  Who knows, maybe tokenizing for blocks is what we need.  I feel like the block handling is the trickiest part as you have to also handle HTML and nested Markdown in HTML.  There may be issues with line handles as well, but this generally area is what needs improvement.

There are probably other minor things I think could be improved, but when I think about the big issues, I think these 2 are probably it.  I did not sit down to write an essay, but it looks like that is how it turned out...

**Comment by facelessuser** [comment]
I just reread your comment, maybe an additional inline parser could be helpful...I'd have to see how things are currently implemented.

**Comment by waylan** [comment]
> Digging back in the code history, I see that the library once had line plugins, and sometimes I think line plugins would actually make handling blocks syntax easier. 

Yeah, there were some things I liked about the 1.x block parser. But it wasn't extendable and was way too recursive (every block went at least another level of recursion---sometime more---so longer documents would reach Python's recursion limit and raise an error). The 2.0 refactor fixed recursion and extendability but added the blank line weirdness because I missed what now seems like an obvious solution (going line by line rather than block by block).

> `"^(.*?)%s(.*?)$" % pattern` seems like an issue which could easily be fixed, but it would not be backwards compatible.

I've never been crazy about that either, but last time I looked into changing it, there were some weird edge cases and I didn't see an obvious way to avoid breaking backward compatibility (the details escape me now). If we are going to do it, now would be the time as we work toward 3.0.

> If I had to pick out Python Markdown's key flaws...

I agree with you on the block parser, but would have to say the second is how nesting is handled for inline patterns. That code came out of one of the GSOC projects we sponsored and while it works good for what it does (and was backward compatible), it is a real headache to maintain and not how anyone expects it to work (bugs almost always result in placeholder strings in the output). To me, the whole `"^(.*?)%s(.*?)$" % pattern` thing seems minor in comparison.

But I think we could address both at once. My thinking regarding the 'two types of inline parsers' solution is that nestable elements (links, bold, italics, etc) should be handled by a tokenizer while non-nestable stuff (code spans, auto links, etc) could be handled by regex. I expect that most user-built extensions would only ever need non-nestable regex (those extensions would be what we have now) or simply override how the tokens are rendered to HTML (each token type should have a replaceable render method). And if an extension really needs to add new tokens, they should be able to add a new type to the list of tokens and then define a matching renderer for that token type.

Given the above, switching to the `regex` lib feels more like a bandaid to me. However, if the barrier to entry isn't to high and it gets us a quick fix for servers timing out from bad input, then that's great. But I don't really see it as a long-term solution to our inline pattern issues. I'me afraid that long term, we are going to need to break backward compatibility and most existing extensions out there will get left behind.

**Comment by facelessuser** [comment]
Okay, the branch is here: https://github.com/facelessuser/markdown/tree/regex.

There were two interesting things I had to change to make it work. Apparently `re` allows you to have a replace function that returns `None`, and it will just replace that `None` with an empty string. I feel like it should have always explicitly returned an empty string, but it is now as it is needed for `regex`: https://github.com/facelessuser/markdown/blob/regex/markdown/inlinepatterns.py#L174.

The other is just a quirk of `regex`, but it honestly makes more sense to me.  We had to change the `hr` pattern from `r'^[ ]{0,3}((-+[ ]{0,2}){3,}|(_+[ ]{0,2}){3,}|(\*+[ ]{0,2}){3,})[ ]*'` to `r'^[ ]{0,3}((-[ ]{0,2}){3,}|(_[ ]{0,2}){3,}|(\*[ ]{0,2}){3,})[ ]*'`. The latter works in Re and Regex, while the former only works in Re.

Everything else worked fine.

> Given the above, switching to the regex lib feels more like a bandaid to me

In general it seems that the tokenizer is a more preferred path, which is also fine. If you are interested in a pull with `regex`, let me know.  I will most likely try and run some 3rd party extension tests to see how some of those do with the regex change.  If you have specific tests in mind, let me know.

I completely understand the desire to move towards a tokenizer in the long term. And if it is ultimately undesirable to switch to `regex` in the short term, no worries. In general I'm driven with this solution as it is a quick path to a solution (and I actually really like regex), but I totally understand if you feel this is not the right solution due it not being pure Python or other reasons.

**Comment by facelessuser** [comment]
Did a little more testing.  I decided to look into the inline efficiency.

So I used the script that you posted over in #161:

```py
# -*-coding:utf8 -*
import markdown
import timeit

txt = '[foo](bar) '*10000
md = markdown.Markdown()

for pname, pattern in md.inlinePatterns.items():
    r = pattern.getCompiledRegExp()
    def test():
        r.match(txt)
    t = timeit.Timer('test()', 'from __main__ import test')
    print ' % 15s:' % pname, t.timeit(number=100)
```

Baseline:

```
        backtick: 2.1914460659
          escape: 0.268756866455
       reference: 1.26284193993
            link: 0.000256061553955
      image_link: 0.26966381073
 image_reference: 0.257260799408
 short_reference: 8.29696655273e-05
        autolink: 0.260588169098
        automail: 0.276794910431
       linebreak: 0.289395093918
            html: 0.314707994461
          entity: 0.310608863831
      not_strong: 0.622181892395
       em_strong: 0.378455877304
       strong_em: 0.381844997406
          strong: 0.809102058411
        emphasis: 0.297507047653
       emphasis2: 0.652032852173
```

But while using recursive regular expression is massively faster at handling **this** issue, it is less efficient at handling `'[foo](bar) '*10000`.  It seems in general `re` is better than `regex` in overall performance, but I guess that makes sense as there are quite a bit more features and logic under the hood of `regex`.

```
        backtick: 3.21650886536
          escape: 0.00902795791626
       reference: 38.5918529034
            link: 0.0353498458862
      image_link: 0.007483959198
 image_reference: 0.00670504570007
 short_reference: 0.000149965286255
        autolink: 0.00876903533936
        automail: 0.00763010978699
       linebreak: 0.0201029777527
            html: 0.00706911087036
          entity: 0.00759792327881
      not_strong: 1.92081713676
       em_strong: 1.14627814293
       strong_em: 1.17105603218
          strong: 1.93952012062
        emphasis: 0.0126938819885
       emphasis2: 0.0104489326477
```

But since I'm in a programming mood, I eliminated `^(.*?)` and `(.*?)$` from inline patterns on [another branch](https://github.com/facelessuser/markdown/tree/inline-pattern-adjust) (it's based on the regex branch as I haven't yet applied it to the master yet).

```
        backtick: 0.000127077102661
          escape: 8.79764556885e-05
       reference: 0.00037693977356
            link: 0.000555992126465
      image_link: 7.20024108887e-05
 image_reference: 5.91278076172e-05
 short_reference: 0.000112056732178
        autolink: 8.48770141602e-05
        automail: 4.91142272949e-05
       linebreak: 4.60147857666e-05
            html: 4.91142272949e-05
          entity: 4.60147857666e-05
      not_strong: 5.00679016113e-05
       em_strong: 4.88758087158e-05
       strong_em: 4.98294830322e-05
          strong: 4.81605529785e-05
        emphasis: 4.6968460083e-05
       emphasis2: 4.60147857666e-05
```

So removing the `^(.*?)` and `(.*?)$` gave us a big speed boost.  It also helped compensate for where recursive regex is weaker.  It would be interesting to do this on a branch without regex as well. I imagine `re` would generally even faster. It really is pretty easy to remove `^(.*?)` and `(.*?)$`, so I think that will be a great move for 3.0.

Regex by itself will fix this current issue, but net performance may not be as good other crazy edge cases, but I think generally will perform good enough.  With more reasonable tests like 1000 links, it handles things quite a bit better.

```
        backtick: 0.324761152267
          escape: 0.0012788772583
       reference: 0.529790878296
            link: 0.000642061233521
      image_link: 0.000814914703369
 image_reference: 0.000685930252075
 short_reference: 0.000133991241455
        autolink: 0.000785112380981
        automail: 0.0013120174408
       linebreak: 0.00296998023987
            html: 0.000680208206177
          entity: 0.000678062438965
      not_strong: 0.193185091019
       em_strong: 0.119333982468
       strong_em: 0.122881889343
          strong: 0.180429935455
        emphasis: 0.000805139541626
       emphasis2: 0.00072193145752
```

Anyways, now we have some data to highlight pros and cons.

**Comment by waylan** [comment]
Interesting results. Thanks for doing that. I see a couple weird anomalies there (`reference: 38.5918529034` being the most obvious). Are those averages of multiple runs, or a single run?

**Comment by facelessuser** [comment]
It isn't specifically an average, but these were fairly repeatable results give or take a little. I'm also on a fairly old macbook.

Yes `reference: 38.5918529034` is clearly high when pushed to an extreme.  I haven't done multiple data points ramping up the amount of data to be processed, so I'm not sure when the performance really drops greatly. Seeing that one thousand is only `reference: 0.529790878296` this a huge jump going to ten thousand.

Now, this is most likely due to recursive patterns for links and references. I imagine replacing the recursive pattern with the same thing we were using with `re`, it would be much closer. While it handles this issue's problem really well, it doesn't seem to be nearly as efficient when pushed in this different extreme case, but maybe there are some things I can do to improve it.

I did expect `re` to be a little faster in some areas mainly because `re` is much lighter in features, and I knew going into this that recursive performance was a bit unknown to me.  I figured it would handle this issue much better, but I wasn't sure about general performance.

With all that said, removing `^(.*?)` and `(.*?)$` was a tremendous  help.  If this test as done anything, it has highlighting how much improvement can still be done in the area of inline handling.

**Comment by facelessuser** [comment]
It is possible we could have two different inline classes: one that inlcudes `^(.*?)` and `(.*?)$`, and one that doesn't.  Or even some kind of alpha flag that tells treeprocessor that the current class is using the new pattern format. This abstraction is doable since all the changes are contained within [`__applyPattern`](https://github.com/facelessuser/markdown/blob/inline-pattern-adjust/markdown/treeprocessors.py#L219).  I would still like to see how this refactor alone affects **this** issue; maybe I'll get some time later today.

**Comment by waylan** [comment]
> It is possible we could have two different inline classes: one that inlcudes ^(.*?) and (.*?)$, and one that doesn't.

This makes for a nice transition. Leave the existing Pattern class as-is and create a new one (perhaps a subclass so that any `isinstance` calls still work) which doesn't  include `^(.*?) and (.*?)$` when `getCompiledRegExp()` gets called. The existing third party extensions would still subclass the old class and continue to work, while the builtin stuff would get the benefits of the new class. Of course, extension authors could then update their extensions to use the new class as they see fit.

**Comment by facelessuser** [comment]
Looks like just transition to an improved inline processor patterns, we gain speed and fix this issue open issue.  So switching to `regex` is not specifically needed.  Yes recursive regular expression allows you handle deeper nested brackets, but it sounds like you want to go to a tokenizer for those kind of things, so we can hold off on something like `regex` for now.

```
        backtick: 6.50882720947e-05
          escape: 0.000121831893921
       reference: 0.000119924545288
            link: 0.000236034393311
      image_link: 5.60283660889e-05
 image_reference: 4.79221343994e-05
 short_reference: 6.38961791992e-05
        autolink: 4.88758087158e-05
        automail: 4.91142272949e-05
       linebreak: 5.00679016113e-05
            html: 4.91142272949e-05
          entity: 4.91142272949e-05
      not_strong: 5.69820404053e-05
       em_strong: 5.10215759277e-05
       strong_em: 4.91142272949e-05
          strong: 5.41210174561e-05
        emphasis: 4.81605529785e-05
       emphasis2: 5.00679016113e-05
```
Just to make clear how much of an improvement we have.  The opening post would hang for a really long time, so modified the test to parse it 10000 times:

```py
# -*-coding:utf8 -*
import markdown
import timeit

txt = 'What I would ultimately like to do is to convert like next;&#xA;[[&quot;03/20/2015&quot;, A, B][&quot;06/03/2015&quot;, C, D][&quot;04/22/2015&quot;, E, F][&quot;03/20/2015&quot;, A2, B2][&quot;06/03/2015&quot;, C2, D2][&quot;01/03/2015&quot;, E2, F2][&quot;12/20/2015&quot;, A3, B3][&quot;11/05/2015&quot;, C3, D3][&quot;07/15/2015&quot;, E3, F3]]&#xA; →&#xA;[[&quot;20/03/2015&quot;, A, B][&quot;03/06/2015&quot;, C, D][&quot;22/04/2015&quot;, E, F][&quot;20/03/2015&quot;, A2, B2][&quot;03/06/2015&quot;, C2, D2][&quot;03/01/2015&quot;, E2, F2][&quot;20/12/2015&quot;, A3, B3][&quot;05/11/2015&quot;, C3, D3][&quot;15/07/2015&quot;, E3, F3]]'*10000
md = markdown.Markdown()

for pname, pattern in md.inlinePatterns.items():
    r = pattern.getCompiledRegExp()
    def test():
        r.match(txt)
    t = timeit.Timer('test()', 'from __main__ import test')
    print ' % 15s:' % pname, t.timeit(number=100)
```

And here are the results:

```
        backtick: 6.89029693604e-05
          escape: 0.000133991241455
       reference: 4.88758087158e-05
            link: 5.00679016113e-05
      image_link: 5.07831573486e-05
 image_reference: 5.19752502441e-05
 short_reference: 4.91142272949e-05
        autolink: 5.10215759277e-05
        automail: 4.88758087158e-05
       linebreak: 4.88758087158e-05
            html: 5.00679016113e-05
          entity: 4.88758087158e-05
      not_strong: 5.79357147217e-05
       em_strong: 5.00679016113e-05
       strong_em: 5.00679016113e-05
          strong: 5.50746917725e-05
        emphasis: 4.91142272949e-05
       emphasis2: 4.98294830322e-05
```

So I still need to abstract the new class.  I'll probably only duplicate the simple base classes, so I plan to duplicate things like `Pattern`, `SimpleTextPattern`, `SimpleTagPattern`, `SubstituteTagPattern`, and `DoubleTagPattern`. Everything else I probably won't duplicate, so if an extension is deriving from something like the `LinkPattern`, that might break, but none of that is documented stuff.  If you feel I shouldn't bother duplicating all of these, or I should duplicate more, let me know.

**Comment by waylan** [comment]
> I'll probably only duplicate the simple base classes, so I plan to duplicate things like `Pattern`, `SimpleTextPattern`, `SimpleTagPattern`, `SubstituteTagPattern`, and `DoubleTagPattern`. Everything else I probably won't duplicate, so if an extension is deriving from something like the `LinkPattern`, that might break, but none of that is documented stuff.  If you feel I shouldn't bother duplicating all of these, or I should duplicate more, let me know.

That sounds reasonable to me.

**Comment by facelessuser** [comment]
Alright, I have a working branch up: https://github.com/facelessuser/markdown/tree/refactor-inline-patterns.  I feel like I should probably create unittests to ensure the now unused, duplicate legacy patterns continue to work, but that is all I feel like doing today.

**Comment by waylan** [comment]
Looks good. Not crazy about the naming though (appending '2' to all the class names). I've always thought the `Pattern` class should be renamed `inlineProcessor` to match with the other parts of the API. Perhaps the new classes should adopt that naming?

**Comment by facelessuser** [comment]
> Not crazy about the naming though (appending '2' to all the class names). I've always thought the Pattern class should be renamed inlineProcessor to match with the other parts of the API. Perhaps the new classes should adopt that naming?

Yeah, I didn't know what to call them, so they got 2 🙂 .  For now I'll leave all the the others as `Pattern`, but the duplicate ones I'll name with `InlineProcessor`.  When we move to 3.0 we can finish renaming the rest.

**Comment by facelessuser** [comment]
Since we are now using the new `InlineProcessor`, I think I may be able to make it more flexible than the old `Pattern` class to allow more flexibility and allow us to handle nested brackets without using regular expression, but still fitting within the same code flow as we currently have that allows regular expression.  Hoping to have a prototype up later today.

**Comment by facelessuser** [comment]
This should allow more flexibility for `InlineProcessor`. This is a proposal, and I'm mainly looking for feedback.  Do we like this approach?  If not, suggestions? Branch is available on the [`flexible-inline` branch](https://github.com/Python-Markdown/markdown/compare/master...facelessuser:flexible-inline).

This branch is built on the [`refactor-inline-patterns`](https://github.com/facelessuser/markdown/tree/refactor-inline-patterns) branch, so if we don't like the approach and can't settle on a new direction, we can just use `refactor-inline-patterns`.

The new branch still performs well with this issue and has the bonus of being able to resolve things too difficult for re with things like tokenizing, or other non regex approaches.  There is still some cleanup to do, and obviously settling on the approach, but all existing unittests pass.

**Comment by waylan** [comment]
> This should allow more flexibility for InlineProcessor. This is a proposal, and I'm mainly looking for feedback. Do we like this approach? If not, suggestions? Branch is available on the flexible-inline branch.

I'm intrigued but not sure I completely follow. Could you update the doc string for [`InlineProcesser.handleMatch`][1] to reflect what should be returned? What are the three items in the tuple and what are valid values for each. How does each of the different valid values change the behavior? 

Also, I get the sense that you anticipate this being useful in ways not used in the existing code. Could you provide an overview of at least one such use case?

[1]: https://github.com/Python-Markdown/markdown/compare/master...facelessuser:flexible-inline#diff-797c6da32780122232df19c7a2e709ebR286

**Comment by facelessuser** [comment]
> I'm intrigued but not sure I completely follow.

I can update the docstring, but I'll try and explain it here.  90% of the time regex is fine. The idea is simply that it takes the text buffer and the current match.  As you see with the link and reference processors, we only use the regex pattern to trigger off the start of the link/ref as we intend to do some more advanced matching than re is capable (particularly recursive brackets). We then use non-regex to handle nested brackets and finish off with a regex pattern for the second part of the link (I'm simply keeping parity with what was already there).

Since we are not solely relying on the initial match object (which contains the start and end of the match), we have to return whatever are new start and end is so we can remove it from the text buffer as has always been done.

Also, since the initial regex match is not sufficient to tell if we have matched what we are looking for, we use `finditer` to iterate through the buffer returning matches, we then let handlematch determine if it is good.  Currently if start and/or end is not valid, we assume that the match was not handled.  We could also return some kind of boolean if that is more clear.

Hopefully that makes sense.

> Also, I get the sense that you anticipate this being useful in ways not used in the existing code. Could you provide an overview of at least one such use case?

I don't know, links and refs is one case.  I can't anticipate if someone will want to to write something that is so "unregular" where regex  may have a difficult time.  This allows a lot more flexibility as shown with links/refs.

**Comment by waylan** [comment]
Very interesting. I think I like it. Once I pulled up the complete file and read it I got it. The partial code in the diff was confusing me because I didn't remember enough about what the missing parts did.

Another thing that would need updated in the doc string is what the `data` argument passed to `handleMatch` is. If I'm not mistaken, its the remaining text from the start of the match to the end of the string being matched.

**Comment by facelessuser** [comment]
I'll definitely document it better moving forward.

Do we want to properly handle recursive `()` in `[links]()`?  This should be really easy to do now, and we can make sure links and images work the same.

**Comment by facelessuser** [comment]
I guess I need to generally see how round brackets are handled currently and in other parsers to be sure how to handle it.

**Comment by waylan** [comment]
> Do we want to properly handle recursive () in [links]()? This should be really easy to do now, and we can make sure links and images work the same.

I would think so. But yes, I'm not sure what "properly" is either without a little research.

**Comment by facelessuser** [comment]
Ugh. I want to disable spaces in href so bad.  It just makes thing more complicated...

I also don't like that a title doesn't need a space between the href.  Some parsers require it, some don't. Looks like Python Markdown doesn't require it.

**Comment by facelessuser** [comment]
Turns out we did have tests for space in URLs from 8 years ago: https://github.com/Python-Markdown/markdown/blob/master/tests/misc/url_spaces.txt.  Not sure how I missed this before; I must have botched my testing.  I'm not sure if that changes are approach to image links (ref #592), if so, now is probably the time.

Anyways, I have some rough code to handle recursive round brackets in links and images with special considerations when in a title (if you are sure you're in a title, they don't have to pair up). There are all sorts of odd cases like `[text](link'"title")` etc. It gets a bit hairy, so I had to make some arbitrary calls, but all existing unittests are passing, and images and links should handle round brackets recursively in their links/files.  I thought at first it was the spaces in href that were a pain, but in the end, I have to say it is the fact that we allow titles to not be spearated from the href with a space.

I'll obviously need to add some new unittests to test for the expectations, but I have something close to a workable solution on the flexible-inline branch.

**Comment by waylan** [comment]
Well, it seems spaces in URLs was not an accident after all, and we have supported them for over ten years how. We probably should be consistent with all URLs (so that includes images). I assume that means that we can inherit the same parsing code in the various subclasses without needing to re-implement things. That would certainly be preferable.

BTW, if you're looking closely at link related tests, it might be a good idea to move them to the new framework in the [utest branch][1]. As a warning, that branch was formed from the now obsolete md3 branch, but the actual changes should cleanly transfer to master. Obviously, doing that would be a separate issue from this, but it makes sense to work on the tests when they are fresh in mind anyway.

[1]: https://github.com/Python-Markdown/markdown/compare/md3...utest

**Comment by facelessuser** [comment]
Yeah, we should be able to reuse all or most of the logic. 

Does the unit test refactor allow legacy tests and new tests to work together? There are a massive amount of tests, and converting them all in one shot will certainly burn me out before I reach the end.

**Comment by waylan** [comment]
> Does the unit test regards allow legacy tests and new tests to work together? There are a massive amount of tests, and converting them all in one shot will certainly burn me out before I reach the end.

Yes, the two can reside side-by-side. I don't expect you to cover everything. But at least the tests you look at. If I have some time this weekend, I'll try to get a new branch started which you could build on.

**Comment by facelessuser** [comment]
Cool. That will be fine then. A path to transition over to new way is perfect.

**Comment by facelessuser** [comment]
I think I finally have some solid defined behavior:

```
Handle nested round brackets.

[Text](http://link.com/(((((((()))))))())) more text

Ensure we close after a title as soon as possible even if brackets are uneven before title.

[Text](http://link.com/("title") more text

[Text](http://link.com/('"title") more text

[Text](http://link.com/("title)") more text

[Text](http://link.com/( "title") more text

[Text](http://link.com/( "title)") more text

We assume the first quote type is what we will match, but we track the second alternate quote type in case we pop the 
last matched bracket before the first type terminates.

[Text](http://link.com/'"title") more text

[Text](http://link.com/"'title') more text

[Text](http://link.com/with spaces'"and quotes" 'and title') more text

[Text](http://link.com/with spaces'"and quotes" 'and title") more text

[Text](http://link.com/with spaces '"and quotes" 'and title') more text

[Text](http://link.com/with spaces "and quotes" 'and title') more text

Angles help with messy quotes.

[Text](<http://link.com/with spaces '"and quotes"> 'and title') more text
```

```html
<p>Handle nested round brackets.</p>
<p><a href="http://link.com/(((((((()))))))())">Text</a> more text</p>
<p>Ensure we close after a title as soon as possible even if brackets are uneven before title.</p>
<p><a href="http://link.com/(" title="title">Text</a> more text</p>
<p><a href="http://link.com/('" title="title">Text</a> more text</p>
<p><a href="http://link.com/(" title="title)">Text</a> more text</p>
<p><a href="http://link.com/(" title="title">Text</a> more text</p>
<p><a href="http://link.com/(" title="title)">Text</a> more text</p>
<p>We assume the first quote type is what we will match, but we track the second alternate quote type in case we pop the 
last matched bracket before the first type terminates.</p>
<p><a href="http://link.com/'" title="title">Text</a> more text</p>
<p><a href="http://link.com/&quot;" title="title">Text</a> more text</p>
<p><a href="http://link.com/with spaces" title="&quot;and quotes&quot; 'and title">Text</a> more text</p>
<p><a href="http://link.com/with spaces'" title="and quotes&quot; 'and title">Text</a> more text</p>
<p><a href="http://link.com/with spaces" title="&quot;and quotes&quot; 'and title">Text</a> more text</p>
<p><a href="http://link.com/with spaces &quot;and quotes&quot;" title="and title">Text</a> more text</p>
<p>Angles help with messy quotes.</p>
<p><a href="http://link.com/with spaces '&quot;and quotes&quot;" title="and title">Text</a> more text</p>
```

**Comment by waylan** [comment]
Cool! And I just pushed up #622 with the new test tools so you can turn all that into new tests. 😉

**Comment by facelessuser** [comment]
Cool. Over the next week I'll see about rebasing off that branch and work out the new tests using the new method.

**Comment by facelessuser** [comment]
I think I'll wait for #622 to get merged first before I rebase.  That way I can be sure I'm rebasing off the final iteration.

**Comment by waylan** [comment]
FYI, #622 has been merged.

**Comment by facelessuser** [comment]
Fantastic! I'll start syncing up my fork to get this pull going.

**Comment by facelessuser** [comment]
Synced up and utilizing new unittest framework.  When pretty smooth.

The last task is unittesting the legacy `Pattern` classes to ensure they don't get broken.  Probably need to double check coverage of new code as well. Hopefully I'll get this done by the end of the week.

**Comment by waylan** [comment]
Fixed in #629.
